io-sim/qsm/Test/StateMachine/IOSim/Execute.hs

```diff
@@ -0,0 +1,740 @@
+                    -- scheduled, i.e. the TQueue writing action was not
+                    -- artificially delayed
+                    fakeInterleaving =
+                        [ not $ hasDoneWork id2 before
+                        | (before, (_, id2, _, EventTxCommitted w _ _) : _) <- events'
+                        , Just "CONTROL" `elem` map l_label w -- has written to the CONTROL TQueue
+                        ]
+                 in counterexample (ppTrace trace) $
+                        -- If this was a fake interleaving just make it pass
+                        property (or fakeInterleaving)
+                            .||. ( fst $
+                                    foldl'
+                                        ( \(prop, model) (cmd, resp) ->
+                                            let model' = transition model cmd resp
+                                                prop' = case logic (postcondition model cmd resp) of
+                                                    VFalse ce -> counterexample ("Poscondition impossible! " <> show ce) $ property False
+                                                    _ -> property True
+                                             in (prop .&&. prop', model')
+                                        )
+                                        (property True, cmodel)
+                                        cmds
+                                 )
+         in
+            ( prevProp .&&. porProp
+            , prevCmds ++ map fst p1' ++ map fst p2'
+            , env1 <> env2
+            , undefined
+            , undefined
+            , foldl (\x (y, z) -> transition x y z) cmodel $ p1' ++ p2'
+            )
+
+    -- This is almost the same as @runSequential@ but it accumulates the concrete
+    -- commands.
+    runSequential' (prevProp, prevCmds, env, smodel, counter, cmodel) (Command scmd _ vars) =
+        let
+            -- Precondition
+            pPre = prec sm smodel scmd
+
+            -- Simulate
+            ccmd = fromRight (error "runSequential: impossible") (reify env scmd)
+            prevCmds' = prevCmds ++ [ccmd]
+            simAction = do
+                sut <- initSut
+                sequence_ $ map (\c -> semantics sut c) prevCmds
+                semantics sut ccmd
+            (trace, cresp, prop) = runIOSim simAction
+
+            -- Postcondition
+            pPost = post sm cmodel ccmd cresp False trace
+            pInv = inv sm cmodel False trace
+
+            -- Mock
+            (pMock, cvars) = mockCheck ccmd cresp vars
+            (sresp, counter') = runGenSym (mock smodel scmd) counter
+         in
+            ( prevProp .&&. pPre .&&. prop .&&. pPost .&&. pInv .&&. pMock
+            , prevCmds'
+            , insertConcretes vars cvars env
+            , transition smodel scmd sresp
+            , counter'
+            , transition cmodel ccmd cresp
+            )
+
+hasDoneWork :: (Eq p) => p -> [(a, p, c, SimEventType)] -> Bool
```

Comment:
Would this work? or maybe `ThreadId m` with a constraint if one needs a polymorphic function.
```suggestion
hasDoneWork :: IOSimThreadId -> [(a, IOSimThreadId, c, SimEventType)] -> Bool
```